packages/legend-studio/src/stores/sidebar-state/WorkspaceWorkflowsState.ts

```diff
@@ -14,35 +14,208 @@
+  treeData.rootIds.push(node.id);
+  treeData.nodes.set(node.id, node);
+  return node;
+};
+
+const addWorkflowJobNodeToTree = (
+  workflowJob: WorkflowJob,
+  workflowNode: WorkflowTreeNodeData,
+  treeData: TreeData<WorkflowExplorerTreeNodeData>,
+): WorkflowJobTreeNodeData => {
+  const node = new WorkflowJobTreeNodeData(workflowJob);
+  if (workflowNode.childrenIds) {
+    workflowNode.childrenIds.push(node.id);
+  } else {
+    workflowNode.childrenIds = [node.id];
+  }
+
+  treeData.nodes.set(node.id, node);
+  return node;
+};
+
+const getWorkflowExplorerTreeNodeData = (
+  workflows: Workflow[],
+): TreeData<WorkflowExplorerTreeNodeData> => {
+  const rootIds: string[] = [];
+  const nodes = new Map<string, WorkflowExplorerTreeNodeData>();
+  const treeData = { rootIds, nodes };
+  workflows.forEach((w) => addWorkflowNodeToTree(w, treeData));
+  return treeData;
+};
+
+const updateWorkflowJobData = (
+  workflowJobs: WorkflowJob[],
+  workflowId: string,
+  treeData: TreeData<WorkflowExplorerTreeNodeData>,
+): void => {
+  const workflowNode = treeData.nodes.get(workflowId);
+  if (workflowNode instanceof WorkflowTreeNodeData) {
+    workflowNode.childrenIds?.forEach((id) => treeData.nodes.delete(id));
+    workflowNode.childrenIds = [];
+    workflowJobs.forEach((job) =>
+      addWorkflowJobNodeToTree(job, workflowNode, treeData),
+    );
+  }
+};
+
+export class WorkflowLogState {
+  editorStore: EditorStore;
+  job: WorkflowJob;
+  logs: string;
+  constructor(editorStore: EditorStore, job: WorkflowJob, logs: string) {
+    makeAutoObservable(this, {
+      editorStore: false,
+      job: observable,
+      logs: observable,
+    });
+
+    this.editorStore = editorStore;
+    this.job = job;
+    this.logs = logs;
+  }
+
+  setLogs(val: string): void {
+    this.logs = val;
+  }
+
+  setJob(val: WorkflowJob): void {
+    this.job = val;
+  }
+
+  *refreshJobLogs(): GeneratorFn<void> {
+    try {
+      const job = WorkflowJob.serialization.fromJson(
+        (yield this.editorStore.sdlcServerClient.getWorkflowJob(
+          this.editorStore.sdlcState.activeProject.projectId,
+          this.editorStore.sdlcState.activeWorkspace,
+          this.job,
+        )) as PlainObject<WorkflowJob>,
+      );
+      const logs = (yield this.editorStore.sdlcServerClient.getWorkflowJobLogs(
+        this.editorStore.sdlcState.activeProject.projectId,
+        this.editorStore.sdlcState.activeWorkspace,
+        this.job,
+      )) as string;
+      this.setJob(job);
+      this.setLogs(logs);
+    } catch (error) {
+      assertErrorThrown(error);
+      this.editorStore.applicationStore.log.error(
+        LogEvent.create(STUDIO_LOG_EVENT.SDLC_MANAGER_FAILURE),
+        error,
+      );
+      this.editorStore.applicationStore.notifyError(error);
+    }
+  }
+}
 
 export class WorkspaceWorkflowsState {
   editorStore: EditorStore;
   sdlcState: EditorSdlcState;
-  isFetchingWorkflows = false;
+  isExecutingWorkflowRequest = false;
```

Comment:
I think this is a limitation we shouldn't have bought. I understand why you're doing this for some of the calls, but for calls like `cancelJob`, `retryJob`, etc. I don't buy it. I think what we should do here is instead of having 1 tree for all the workflows, have 1 tree per workflow, and have that a wrapper state for it, something like `WorkspaceWorkflowState` which has a flag `isExecutiongWorkflowRequest` on it and a tree state. On `WorkspaceWorkflowsState`, we can have different `ActionState` for things like `fetchingWorkflows`, etc. individually. Of course, one small setback is that we have to keep `workflowJobLogState` here in `WorkspaceWorkflowsState` as we don't want to allow stacking multiple modals.

Reply:
Don't you still have the same problem with your approach. Where would you put the flag for cancelling/retrying a specific job. Unless we go and create yet another state for each individual job which seems a bit excessive.
Is it wrong just to create another flag called `isExecutingWorkflowJobRequest`. The only purpose of this flag is to alert the global workspace workflow state of a loading state.

Reply:
`isExecutingWorkflowJobRequest` in `WorkspaceWorkflowState` is totally fine. But job operations' state should be local to the workflow, not at `WorkspaceWorkflowsState` level.